theano/tensor/nnet/nnet.py

```diff
@@ -2215,8 +2215,7 @@ def relu(x, alpha=0):
         return f1 * x + f2 * abs(x)
 
 
-def h_softmax(x, batch_size, n_outputs, n_classes, n_outputs_per_class,
-              W1, b1, W2, b2, target=None):
+def h_softmax(x, n_outputs, W1, b1, W2, b2, target=None):
```

Comment:
> This break the interface... We don't like that. @adbrebs and @bouthilx What do you think of that? I saw one commit that changed the size to be in front instead of later parameter, what was the reason? Why you added those n_* that can be computed?

The parameters `n_classes` and `n_outputs_per_class` were optional as we were trying to deduce them from `n_outputs` . We draw back as the shapes are implied by `W1` and `W2`. See discussion [here](https://github.com/Theano/Theano/pull/3361#discussion_r41105999). We moved them to the front as they were not optional anymore, thus grouping size parameters together in `h_softmax` function's signature.

I believe the `n_*` parameters might just be an artifact from groundhog's [code](https://github.com/lisa-groundhog/GroundHog/blob/master/groundhog/layers/cost_layers.py#L1163). They create weights matrices using given size parameters, and then used the size parameters directly to compute sparse_block_dot rather than inferring them from weights. It seems we could indeed infer from weights rather than using explicit size parameters as we do not create weights and get it as arguments.

To me the best signature would be 
```python
h_softmax(x, W1, b1, W2, b2, n_outputs, batch_size=None, n_outputs=None, n_classes=None, n_outputs_per_class=None, target=None).
```
Otherwise, I prefer solution 3).